src/libmongoc/tests/unified/operation.c

```diff
@@ -0,0 +1,2259 @@
+/*
+ * Copyright 2020-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "operation.h"
+
+#include "mongoc-array-private.h"
+#include "result.h"
+#include "test-diagnostics.h"
+#include "util.h"
+#include "utlist.h"
+
+typedef struct {
+   char *name;
+   char *object;
+   bson_t *arguments;
+   bson_t *expect_error;
+   bson_val_t *expect_result;
+   char *save_result_as_entity;
+   bson_parser_t *parser;
+   char *session_id;
+   mongoc_client_session_t *session;
+} operation_t;
+
+static void
+operation_destroy (operation_t *op)
+{
+   if (!op) {
+      return;
+   }
+   bson_parser_destroy_with_parsed_fields (op->parser);
+   bson_free (op->session_id);
+   bson_free (op);
+}
+
+static operation_t *
+operation_new (bson_t *bson, bson_error_t *error)
+{
+   operation_t *op = bson_malloc0 (sizeof (operation_t));
+   op->parser = bson_parser_new ();
+   bson_parser_utf8 (op->parser, "name", &op->name);
+   bson_parser_utf8 (op->parser, "object", &op->object);
+   bson_parser_doc_optional (op->parser, "arguments", &op->arguments);
+   bson_parser_doc_optional (op->parser, "expectError", &op->expect_error);
+   bson_parser_any_optional (op->parser, "expectResult", &op->expect_result);
+   bson_parser_utf8_optional (
+      op->parser, "saveResultAsEntity", &op->save_result_as_entity);
+   if (!bson_parser_parse (op->parser, bson, error)) {
+      operation_destroy (op);
+      return NULL;
+   }
+   return op;
+}
+
+static bool
+operation_create_change_stream (test_t *test,
+                                operation_t *op,
+                                result_t *result,
+                                bson_error_t *error)
+{
+   bool ret = false;
+   entity_t *entity = NULL;
+   bson_parser_t *parser = NULL;
+   mongoc_change_stream_t *changestream = NULL;
+   bson_t *pipeline = NULL;
+   const bson_t *op_reply = NULL;
+   bson_error_t op_error = {0};
+
+   parser = bson_parser_new ();
+   /* Capture options as all extra fields, and pass them directly as change
+    * stream options. */
```

Comment:
The rationale for passing extra options along is to avoid ignoring any future options added to tests.

Operations in libmongoc pass extra options along to commands. That will result in a server error, so it should result in a test failure.